### PR TITLE
Model side tracking and remove unnecessary base model interface if not needed

### LINKED ISSUE #160: Implement model-side tracking
Trackers could be placed in the model to make them compatible even with wrapped attacks.

### LINKED ISSUE #195: The model wrapper can be made optional
Make it possible to remove this line:

```diff
model = ...
torch_data_loader = ...

- # Wrap model
- model = BasePyTorchClassifier(model)

# create and run attack
attack = PGD(
    perturbation_model="l2",
    epsilon=0.4,
    num_steps=100,
    step_size=0.01,
)

adversarial_loader = attack(model, torch_data_loader)

# Test accuracy on adversarial examples
robust_accuracy = Accuracy()(model, adversarial_loader)
```

The wrapper can be instantiated inside the attack and metrics if needed. This has to be optional, as many models actually need differentiable chainable pre- and post-processing modules.

### PR DESCRIPTION
**Pull request by maurapintor**
Several improvements and updates to the example scripts and the attack classes. The most significant changes are **the removal of unnecessary `BasePyTorchClassifier` wrappers if not needed**, the addition of new and improved trackers enabling support for other backends (AdvLib and Foolbox). 

**Removed BasePyTorchClassifier mandatory wrapping:**
The wrapper is still supported if needed (e.g., in case of preprocessing or postprocessing operations).

**Example simplification and consistency:**

* Removed the use of `BasePyTorchClassifier` in all example scripts, passing raw `nn.Module` models directly to training, evaluation, and attack methods for a simpler and more consistent API. 

**New and improved adversarial attack examples:**

* Added new example scripts: `boundary_attack_example.py` demonstrating the Boundary Attack with Foolbox and tracker integration, and `model_tracker_example.py` showing how to use model-level tracking with different attack backends.
* Updated `mnist_example.py` and `evasion_fmn_example.py` to support trackers for Foolbox and AdvLib backends, and to demonstrate robust accuracy evaluation across multiple attack implementations. 

**Tracker and backend support improvements:**

* Enhanced tracker integration: Trackers such as `LossTracker`, `PredictionTracker`, `PerturbationNormTracker`, and `GradientNormTracker` can now be passed to attacks using native, Foolbox, and AdvLib backends in the examples. 
* Enabled tracker support for AdvLib attacks by updating the `BaseAdvLibEvasionAttack` class to accept and allow trackers, aligning its interface with other backends. 

**Other improvements:**

* Updated attack hyperparameters and model weights in examples (e.g., using "teacher" weights, adjusting `epsilon` and `num_steps`).

### CODE REVIEW

**Comment by maurapintor** [comment]
This is an example of plot made with the three backends with the newly-implemented trackers:
image

**Comment by maurapintor** [comment]
@zangobot this is ready for review. It might be long to review as it touches many files for two separate changes and some minor fixes in the docs. Let me know if you need to sync about this and look at it together.

**Review by zangobot — APPROVED** [review]
LGTM!
